src/OpenTelemetry.Instrumentation.Wcf/Implementation/HttpRequestMessagePropertyWrapper.cs

```diff
@@ -0,0 +1,111 @@
+/// will be disabled (IsHttpFunctionalityEnabled == false).
+/// </summary>
+internal class HttpRequestMessagePropertyWrapper
+{
+    private static readonly ReflectedInfo ReflectedValues = Initialize();
+    private readonly object innerObject;
+
+    public HttpRequestMessagePropertyWrapper(object existingHttpRequestMessageProperty = null)
+    {
+        EnsureEnabled();
+        if (existingHttpRequestMessageProperty != null && !existingHttpRequestMessageProperty.GetType().Equals(ReflectedValues.Type))
+        {
+            throw new ArgumentException("Existing object must be of type HttpRequestMessageProperty", nameof(existingHttpRequestMessageProperty));
+        }
+
+        this.innerObject = existingHttpRequestMessageProperty != null ? existingHttpRequestMessageProperty : Activator.CreateInstance(ReflectedValues.Type);
+    }
+
+    public static bool IsHttpFunctionalityEnabled => ReflectedValues != null;
+
+    public static string Name
+    {
+        get
+        {
+            EnsureEnabled();
+            return ReflectedValues.Name;
+        }
+    }
+
+    public WebHeaderCollection Headers => ReflectedValues.HeadersFetcher.Fetch(this.innerObject);
+
+    public object InnerObject => this.innerObject;
+
+    private static ReflectedInfo Initialize()
+    {
+        Type type = null;
+        try
+        {
+            type = Type.GetType(
+                "System.ServiceModel.Channels.HttpRequestMessageProperty, System.ServiceModel, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
+                true);
+
+            var headersProp = type.GetProperty("Headers", BindingFlags.Public | BindingFlags.Instance, null, typeof(WebHeaderCollection), Array.Empty<Type>(), null);
+            if (headersProp == null)
+            {
+                throw new NotSupportedException("HttpRequestMessageProperty.Headers property not found");
+            }
+
+            var nameProp = type.GetProperty("Name", BindingFlags.Public | BindingFlags.Static, null, typeof(string), Array.Empty<Type>(), null);
+            if (nameProp == null)
+            {
+                throw new NotSupportedException("HttpRequestMessageProperty.Name property not found");
+            }
+
+            return new ReflectedInfo
+            {
+                Type = type,
+                Name = (string)nameProp.GetValue(null),
+                HeadersFetcher = new PropertyFetcher<WebHeaderCollection>(nameof(Headers)),
+            };
+        }
+        catch (Exception ex)
+        {
+            WcfInstrumentationEventSource.Log.HttpServiceModelReflectionFailedToBind(ex, type?.Assembly);
+        }
+
+        return null;
+    }
+
+    private static void EnsureEnabled()
+    {
+        if (!IsHttpFunctionalityEnabled)
+        {
+            throw new InvalidOperationException("Http functionality is not enabled, check IsHttpFunctionalityEnabled before calling this method");
+        }
+    }
```

Comment:
It seems like the calling code on the PR is checking `IsHttpFunctionalityEnabled` correctly so is the goal here to catch potential mistakes at dev time? Could we move under a debug block maybe?

```csharp
    private static void EnsureEnabled()
    {
#if DEBUG
        if (!IsHttpFunctionalityEnabled)
        {
            throw new InvalidOperationException("Http functionality is not enabled, check IsHttpFunctionalityEnabled before calling this method");
        }
#endif
    }
```

Reply:
I've refactored this safety check and the object type check to be DEBUG-only 👍